File: companion/bolumler/b06/ornek-01/cozum.R
hazirla <- function(veri) {
  if (!identical(names(veri), c("id", "sinif", "basit", "tabakali")) || nrow(veri) != 12) {
    stop("Bu pilot 12 kisilik cerceveyi bekler.")
  }
  if (!all(vapply(veri, is.numeric, logical(1))) || anyNA(veri) ||
      any(!is.finite(as.matrix(veri)))) stop("Sonlu ve eksiksiz sayisal degerler gerekli.")
  temiz <- veri[order(veri$id), ]
  if (!all(temiz$id == 1:12) || !all(temiz$sinif == rep(1:3, each = 4))) {
    stop("Kimlik-sinif eslesmesi kitap cercevesiyle uyusmuyor.")
  }
  if (!all(as.matrix(temiz[c("basit", "tabakali")]) %in% c(0, 1))) stop("Gostergeler 0/1 olmali.")
  if (sum(temiz$basit) != 6 || !all(tapply(temiz$tabakali, temiz$sinif, sum) == 2)) {
    stop("Orneklem buyuklugu veya tabaka tahsisi uyusmuyor.")
  }
  temiz$sinif <- factor(temiz$sinif, levels = 1:3, ordered = TRUE)
  temiz
}

hesapla <- function(veri) {
  temiz <- hazirla(veri)
  sonuc <- data.frame(degisken = rep("cerceve", 3),
                      olcu = c("hacim", "benzersiz_id", "sinif_sayisi"),
                      deger = c(nrow(temiz), length(unique(temiz$id)), nlevels(temiz$sinif)),
                      stringsAsFactors = FALSE)
  for (tasarim in c("basit", "tabakali")) {
    secilen <- temiz[temiz[[tasarim]] == 1, ]
    olasilik <- if (tasarim == "basit") nrow(secilen) / nrow(temiz) else 2 / 4
    agirlik <- 1 / olasilik
    sonuc <- rbind(sonuc, data.frame(degisken = rep(tasarim, 13),
      olcu = c("orneklem_hacmi", "dahil_edilme_olasiligi", "tasarim_agirligi", "agirlik_toplami",
               paste0("sinif_", 1:3, "_frekans"), paste0("secilen_id_", 1:6)),
      deger = c(nrow(secilen), olasilik, agirlik, nrow(secilen) * agirlik,
                as.numeric(table(secilen$sinif)), secilen$id)))
  }
  sonuc
}

kontrol_et <- function(sonuc, beklenen) {
  if (!identical(names(beklenen), c("degisken", "olcu", "deger")) ||
      !identical(sonuc$degisken, beklenen$degisken) ||
      !identical(sonuc$olcu, beklenen$olcu)) stop("Kontrol etiketleri uyusmuyor.")
  if (!is.numeric(beklenen$deger) || anyNA(beklenen$deger) ||
      any(!is.finite(beklenen$deger)) || any(!is.finite(sonuc$deger))) stop("Sonlu kontrol gerekli.")
  tolerans <- pmax(1e-9, 1e-9 * pmax(abs(sonuc$deger), abs(beklenen$deger)))
  if (any(abs(sonuc$deger - beklenen$deger) > tolerans)) stop("Kontrol uyusmazligi.")
  cat("DOGRULANDI:", nrow(sonuc), "kontrol degeri eslesiyor.\n")
}

grafik_kaydet <- function(veri) {
  temiz <- hazirla(veri)
  dir.create("ciktilar/r", recursive = TRUE, showWarnings = FALSE)
  png("ciktilar/r/secim-haritasi.png", width = 1500, height = 525, res = 150)
  on.exit(dev.off())
  par(mar = c(4, 7, 3, 1))
  plot(NA, xlim = c(0.5, 12.5), ylim = c(-0.5, 2.8), xaxt = "n", yaxt = "n",
       xlab = "Yapay ogrenci kimligi", ylab = "", main = "Kayitli secimler: dolu nokta secilen birim")
  axis(1, at = 1:12)
  axis(2, at = 0:2, labels = c("Tabakali", "Basit rastgele", "Cerceve"), las = 1)
  abline(v = c(4.5, 8.5), lty = 3, col = "grey")
  text(c(2.5, 6.5, 10.5), 2.4, paste("Sinif", 1:3))
  points(temiz$id, rep(2, 12), pch = 19, col = "#555555")
  for (index in 1:2) {
    tasarim <- c("basit", "tabakali")[index]
    konum <- 2 - index
    points(temiz$id, rep(konum, 12), pch = 1, col = "grey")
    secilen <- temiz$id[temiz[[tasarim]] == 1]
    points(secilen, rep(konum, length(secilen)), pch = 19, col = c("#1F4E79", "#087F5B")[index])
  }
  cat("Grafik: ciktilar/r/secim-haritasi.png\n")
}

secenek <- commandArgs(trailingOnly = TRUE)
if (any(!secenek %in% c("--check", "--grafik"))) stop("Bilinmeyen secenek.")
veri <- read.csv("veri.csv", check.names = FALSE)
sonuc <- hesapla(veri)
print(sonuc, row.names = FALSE)
if ("--check" %in% secenek) kontrol_et(sonuc, read.csv("beklenen-sonuclar.csv", stringsAsFactors = FALSE))
if ("--grafik" %in% secenek) grafik_kaydet(veri)
sessionInfo()

File: companion/bolumler/b06/ornek-01/cozum.py
import argparse
import math
from pathlib import Path
import platform

import numpy as np
import pandas as pd


def hazirla(veri):
    if list(veri.columns) != ["id", "sinif", "basit", "tabakali"] or len(veri) != 12:
        raise ValueError("Bu pilot dort sutunlu, 12 kisilik cerceveyi bekler.")
    if veri.isna().any().any():
        raise ValueError("Eksik deger kabul edilmez.")
    for sutun in veri.columns:
        if not pd.api.types.is_numeric_dtype(veri[sutun]) or pd.api.types.is_bool_dtype(veri[sutun]):
            raise ValueError("Sayisal sutunlar gerekli.")
    if not np.isfinite(veri.to_numpy(dtype=float)).all():
        raise ValueError("Sonlu degerler gerekli.")
    temiz = veri.sort_values("id").reset_index(drop=True).copy(deep=True)
    if not np.array_equal(temiz["id"], np.arange(1, 13)):
        raise ValueError("Kimlikler 1..12 arasinda benzersiz olmali.")
    if not np.array_equal(temiz["sinif"], np.repeat([1, 2, 3], 4)):
        raise ValueError("Kimlik-sinif eslesmesi kitap cercevesiyle uyusmuyor.")
    if not temiz[["basit", "tabakali"]].isin([0, 1]).all().all():
        raise ValueError("Secim gostergeleri 0 veya 1 olmali.")
    if temiz["basit"].sum() != 6:
        raise ValueError("Basit secimde alti farkli birim gerekli.")
    adetler = temiz.groupby("sinif")["tabakali"].sum()
    if not (adetler == 2).all():
        raise ValueError("Tabakali secimde her siniftan iki birim gerekli.")
    temiz = temiz.astype(int)
    temiz["sinif"] = pd.Categorical(temiz["sinif"], categories=[1, 2, 3], ordered=True)
    return temiz


def hesapla(veri):
    temiz = hazirla(veri)
    satirlar = [("cerceve", "hacim", len(temiz)),
                ("cerceve", "benzersiz_id", temiz["id"].nunique()),
                ("cerceve", "sinif_sayisi", temiz["sinif"].nunique())]
    for tasarim in ["basit", "tabakali"]:
        secilen = temiz[temiz[tasarim] == 1]
        olasilik = len(secilen) / len(temiz) if tasarim == "basit" else 2 / 4
        agirlik = 1 / olasilik
        satirlar.extend([(tasarim, "orneklem_hacmi", len(secilen)),
                         (tasarim, "dahil_edilme_olasiligi", olasilik),
                         (tasarim, "tasarim_agirligi", agirlik),
                         (tasarim, "agirlik_toplami", len(secilen) * agirlik)])
        for sinif in [1, 2, 3]:
            satirlar.append((tasarim, f"sinif_{sinif}_frekans", (secilen["sinif"] == sinif).sum()))
        satirlar.extend((tasarim, f"secilen_id_{sira}", kimlik)
                        for sira, kimlik in enumerate(secilen["id"], start=1))
    return pd.DataFrame(satirlar, columns=["degisken", "olcu", "deger"])


def kontrol_et(sonuc, beklenen):
    if list(beklenen.columns) != ["degisken", "olcu", "deger"]:
        raise ValueError("Kontrol sutunlari uyusmuyor.")
    if not sonuc[["degisken", "olcu"]].equals(beklenen[["degisken", "olcu"]]):
        raise ValueError("Kontrol etiketleri veya sirasi uyusmuyor.")
    hedefler = pd.to_numeric(beklenen["deger"], errors="raise")
    if not np.isfinite(hedefler).all() or not np.isfinite(sonuc["deger"]).all():
        raise ValueError("Sonlu kontrol degerleri gerekli.")
    for gercek, hedef in zip(sonuc["deger"], hedefler):
        if not math.isclose(gercek, hedef, rel_tol=1e-9, abs_tol=1e-9):
            raise ValueError(f"Kontrol uyusmazligi: {gercek} != {hedef}")
    print(f"DOGRULANDI: {len(sonuc)} kontrol degeri eslesiyor.")


def grafik_kaydet(veri):
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    temiz = hazirla(veri)
    fig, eksen = plt.subplots(figsize=(10, 3.5))
    for konum, tasarim, renk in [(2, None, "#555555"), (1, "basit", "#1F4E79"), (0, "tabakali", "#087F5B")]:
        eksen.scatter(temiz["id"], [konum] * 12, s=120, facecolors="white", edgecolors="#AAAAAA")
        secilen = temiz if tasarim is None else temiz[temiz[tasarim] == 1]
        eksen.scatter(secilen["id"], [konum] * len(secilen), s=120, color=renk)
    for sinir in [4.5, 8.5]:
        eksen.axvline(sinir, color="#BBBBBB", linestyle=":")
    for merkez, sinif in [(2.5, 1), (6.5, 2), (10.5, 3)]:
        eksen.text(merkez, 2.4, f"Sinif {sinif}", ha="center")
    eksen.set(xlim=(0.5, 12.5), ylim=(-0.5, 2.8), xticks=range(1, 13),
              yticks=[0, 1, 2], yticklabels=["Tabakali", "Basit rastgele", "Cerceve"],
              xlabel="Yapay ogrenci kimligi", title="Kayitli secimler: dolu nokta secilen birimi gosterir")
    fig.tight_layout()
    hedef = Path("ciktilar/python")
    hedef.mkdir(parents=True, exist_ok=True)
    fig.savefig(hedef / "secim-haritasi.png", dpi=150)
    plt.close(fig)
    print("Grafik: ciktilar/python/secim-haritasi.png")


def main():
    parser = argparse.ArgumentParser(description="B06: kayitli basit ve tabakali secimleri cozumle")
    parser.add_argument("--check", action="store_true")
    parser.add_argument("--grafik", action="store_true")
    secenek = parser.parse_args()
    veri = pd.read_csv("veri.csv")
    sonuc = hesapla(veri)
    print(sonuc.to_string(index=False))
    if secenek.check:
        kontrol_et(sonuc, pd.read_csv("beklenen-sonuclar.csv"))
    if secenek.grafik:
        grafik_kaydet(veri)
    print(f"Python {platform.python_version()}, NumPy {np.__version__}, pandas {pd.__version__}")


if __name__ == "__main__":
    main()